src/main/java/org/intellij/sdk/codesync/codeSyncSetup/S3FileUploader.java

```diff
@@ -0,0 +1,133 @@
+
+public class S3FileUploader {
+    S3UploadQueueFile s3UploadQueueFile;
+    Integer runCount;
+    Map<String, Object> failedFilePathsAndURLs = new HashMap<>();
+
+    final Integer maxRunCount = 10;
+
+    public S3FileUploader (S3UploadQueueFile s3UploadQueueFile) {
+        this.s3UploadQueueFile = s3UploadQueueFile;
+        this.runCount = this.s3UploadQueueFile.getRunCount();
+    }
+    public S3FileUploader (String repoPath, String branch, Map<String, Object> filePathAndURLs) {
+        this.runCount = 0;
+        this.s3UploadQueueFile = new S3UploadQueueFile(repoPath, branch, filePathAndURLs, runCount);
+    }
+
+    public void saveURLs() throws InvalidYmlFileError, FileNotFoundException {
+        try {
+            this.s3UploadQueueFile.writeYml();
+        } catch (FileLockedError error) {
+            // This would never happen, but adding it for completeness.
+            CodeSyncLogger.error("File locked error while creating a new file for S3 file upload queue.");
+        }
+    }
+
+    /*
+    Pre-Process and validate yml data. return `true` of we are good to proceed, `false` otherwise.
+    */
+    public boolean isValid() {
+        if (!this.s3UploadQueueFile.hasFiles()){
+            return false;
+        }
+
+        // If previous 3 attempts have resulted in failures then do not try again.
+        return this.runCount <= this.maxRunCount;
+    }
+
+    public void uploadToS3(String repoPath, String branchName, Map<String, Object> fileUrls) {
+        OriginalsRepoManager originalsRepoManager = new OriginalsRepoManager(repoPath, branchName);
+        CodeSyncClient codeSyncClient = new CodeSyncClient();
+
+        for (Map.Entry<String, Object> fileUrl : fileUrls.entrySet()) {
+            File originalsFile = originalsRepoManager.getFilePath(fileUrl.getKey()).toFile();
+            if (fileUrl.getValue().equals("")) {
+                // Skip if file is empty.
+                continue;
+            }
+            try {
+                codeSyncClient.uploadToS3(originalsFile, (Map<String, Object>) fileUrl.getValue());
+
+                // File uploaded with success, so delete it from the originals.
+                originalsRepoManager.deleteFile(originalsFile.getPath());
+            } catch (RequestError error) {
+                this.failedFilePathsAndURLs.put(fileUrl.getKey(), fileUrl.getValue());
+                CodeSyncLogger.critical(
+                    String.format("[S3_FILE_UPLOAD]: Could not upload file to S3. Error %s", error.getMessage())
+                );
+            } catch (ClassCastException error) {
+                CodeSyncLogger.critical(
+                    String.format("[S3_FILE_UPLOAD]: Could not process yaml content. Error %s", error.getMessage())
+                );
+            }
+        }
+
+        if (this.failedFilePathsAndURLs.isEmpty()) {
+            // Remove Originals repo as they are not needed anymore.
+            originalsRepoManager.delete();
```

Comment:
@Basit-qc I think we should keep this in place to remove all the unneeded files that could have been persisted for any reason.
I do not think we would have multiple queue files for the same repo.

Reply:
Plugin uploads new files individually via `/v1/files` which returns preSignedURL. We should also save that as Queue file so that PreSignedURL is not lost after the API call. 
Ref from VSCode: https://github.com/codesyncapp/codesync-vs/blob/master/src/utils/upload_utils.ts#L189

That being said, s3Uploader will be uploading single files individually for 1 repo, so e.g. 
1. `git pull` results in X new files OR
2. Since `/v1/init` processes only 800 files, so rest of the files are uploaded individually

So there will be multiple queue files at 1 time. Uploading those files should work fine. If we delete `.originals` repo after 1 successful upload, all originals files will be lost.

@saleem-latif let me know your thoughts

Reply:
Thanks for clarifying this, makes sense, I have removed this part.